mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,563 @@
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKGEMMOUTPUTSWIZZLEPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-gemm-output-swizzle"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace mlir::rock;
+using mlir::gpu::AddressSpace;
+
+namespace {
+struct RockGemmOutputSwizzlePass
+    : public rock::impl::RockGemmOutputSwizzlePassBase<
+          RockGemmOutputSwizzlePass> {
+  void runOnOperation() override;
+};
+
+bool hasPrivateMemoryAddressSpace(MemRefType type) {
+  Attribute memorySpace = type.getMemorySpace();
+  if (!memorySpace)
+    return false;
+  if (auto gpuAttr = llvm::dyn_cast<gpu::AddressSpaceAttr>(memorySpace)) {
+
+    return gpuAttr.getValue() == AddressSpace::Private;
+  }
+  return false;
+}
+
+bool hasWorkgroupMemoryAddressSpace(MemRefType type) {
+  Attribute memorySpace = type.getMemorySpace();
+  if (!memorySpace)
+    return false;
+  if (auto gpuAttr = llvm::dyn_cast<gpu::AddressSpaceAttr>(memorySpace)) {
+
+    return gpuAttr.getValue() == AddressSpace::Workgroup;
+  }
+  return false;
+}
+
+bool hasGlobalMemoryAddressSpace(MemRefType type) {
+  return !hasWorkgroupMemoryAddressSpace(type) &&
+         !hasPrivateMemoryAddressSpace(type);
+}
+
+int64_t getLDSTotalSize(func::FuncOp &func) {
+  int64_t totalSize = 0;
+  func.walk([&](GpuAllocOp gpuAlloc) {
+    auto type = gpuAlloc.getOutput().getType();
+    auto memSpaceValue =
+        dyn_cast_or_null<gpu::AddressSpaceAttr>(type.getMemorySpace())
+            .getValue();
+    if (memSpaceValue == gpu::GPUDialect::getWorkgroupAddressSpace()) {
+      totalSize += type.getNumElements() * getByteWidth(type.getElementType());
+    }
+  });
+  return totalSize;
+}
+
+static LogicalResult checkLDSSize(Operation *op, int64_t ldsBytes) {
+  // Check for arch limitations exceeded
+  FailureOr<StringAttr> maybeArch = getArch(op);
+  if (succeeded(maybeArch)) {
+    StringAttr arch = maybeArch.value();
+    const int64_t ldsSize = rock::lookupArchInfo(arch).maxSharedMemPerWG;
+    return success(ldsBytes <= ldsSize);
+  }
+  return success();
+}
+
+// Function to create the schedule of the current set of stages
+void reuseDeadLDS(func::FuncOp &func) {
+  // Retrieve all GpuAlloc operations targeting LDS and
+  // calculate the total memory space required for all operations
+  // except the last one. Additionally, compute the memory space
+  // required by the last GpuAlloc operation.
+  SmallVector<GpuAllocOp> allocs;
+  int64_t totalSize = 0;
+  int64_t lastSize = 0;
+  func.walk([&](GpuAllocOp gpuAlloc) {
+    auto type = gpuAlloc.getOutput().getType();
```

Comment:
Might be good to spell out `MemRefType` here - upstream style (which we broadly try to hold to) says to use auto only in cases where the type is either obvious from context (like in those `dyn_cast<>`s below) or long and pointless to name (ex. various iterators)